src/sentry/notifications/notifications/base.py

```diff
@@ -85,6 +85,19 @@ def format_url(self, text: str, url: str) -> str:
         """
         return self.url_format.format(text=text, url=url)
 
+    @property
+    def provider(self) -> ExternalProviders:
+        if not getattr(self, "_provider", None):
+            raise AttributeError(
+                f"'provider' not set on {self.__class__.__name__}. Please set 'provider' from the message builder."
+            )
+
+        return self._provider
+
+    @provider.setter
+    def provider(self, provider: ExternalProviders) -> None:
+        self._provider = provider
```

Comment:
@97varun I'm confused why we were making `._provider` stateful like this. This has the potential of introducing many bugs because a single instance of a notification is supposed to work for many different providers. We can't just store state of the last provider, we need to pass it in explicitly wherever we need it

Reply:
@scefali We don't mutate the object instead, we're creating a new object with these properties.
https://github.com/getsentry/sentry/blob/8d642af13fe3c323bfcb660db0dc54263a41d852/src/sentry/integrations/slack/message_builder/issues.py#L350-L353

I agree that this is not great, but passing in parameters everywhere for both `provider` and `url_format` didn't feel right to me as sometimes these are used a level deeper, so the parameters have to be cascaded. Like `get_description_as_text` and `get_notification_title` here.

https://github.com/getsentry/sentry/blob/8d642af13fe3c323bfcb660db0dc54263a41d852/src/sentry/notifications/notifications/activity/base.py#L134-L136

https://github.com/getsentry/sentry/blob/8d642af13fe3c323bfcb660db0dc54263a41d852/src/sentry/notifications/notifications/activity/base.py#L141-L158

Ideally I would like to create new types and build new objects using a constructor for each provider, but by the time we know which provider it is, we get the instantiated object, and it would be slightly complicated to construct a new object out of it. 

Please let me know if you still think it is better to do this in a particular way, and I'll make the changes.

Reply:
@97varun I see in this case you make a new notification, but that's not how it always works. Ee: https://github.com/getsentry/sentry/blob/master/src/sentry/notifications/notifications/base.py#L243-L245

So we cannot bind the provider to the notification

Reply:
Updated to pass an argument for `provider`.